src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/core/message_infrastructure_logging.h

```diff
@@ -5,39 +5,157 @@
 #ifndef MESSAGE_INFRASTRUCTURE_LOGGING_H_
 #define MESSAGE_INFRASTRUCTURE_LOGGING_H_
 
+#include <memory>
+#include <mutex>
+#include <thread>
+#include <iostream> 
+#include <sstream> 
+#include <fstream>
+#include <string>
+#include <cstdlib>
+
 #include <stdio.h>
-#define LOG_MP (0)  // log for multiprocessing
-#define LOG_ACTOR (0)
-#define LOG_LAYER (0)
+#include <memory.h>
+#include <time.h>
+
+#if _WIN32
+#inlcude <process.h>
+#include <direct.h> // _getcwd()
+#define getpid() _getpid()
+#define getcwd() _getcwd()
+#else //__linux__ & __APPLE__
+#include <unistd.h>
+#include <sys/types.h>
+#endif
+
+#ifdef CMAKE_DEBUG
+#define DEBUG_MODE (1)
+#else
 #define DEBUG_MODE (0)
-#define LOG_SMMP (0)  // log for shmemport
+#endif
+
+#define DEBUG_LOG_MODULE "lava_message_infrastructure"
+#define DEBUG_LOG_FILE_SUFFIX "log"
+#define NULL_STRING ""
+
+//#if DEBUG_MODE
+#define DEBUG_LOG_TO_FILE (1)
+
+#define MAX_SIZE_PER_LOG_MSG (1024)
+#define MAX_SIZE_LOG (1)
+#define MAX_SIZE_LOG_TIME (64)
 
-#define LAVA_LOG(_cond, _fmt, ...) { \
-  if ((_cond)) { \
-    printf("[CPP INFO]" _fmt, ## __VA_ARGS__); \
+#define LOG_MSG_SUBSTITUTION "!This message was displayed due to the failure of the malloc of this log message!\n"
+#define LOG_GET_TIME_FAIL "Get log time failed."
+
+#if DEBUG_MODE && DEBUG_LOG_TO_FILE
+#define DEBUG_LOG_PRINT(_level,_module, _fmt, ...) do {\
+  int length = 0;\
+  char * log_data = (char*)malloc(sizeof(char)*MAX_SIZE_PER_LOG_MSG);\
+  if (log_data != NULL)\
+    length = std::snprintf(log_data, MAX_SIZE_PER_LOG_MSG, _fmt,  ## __VA_ARGS__);\
+  if (log_data == NULL || length <0){\
+    getLogInstance()->log(new LogMsg(std::string(LOG_MSG_SUBSTITUTION), __FILE__, __LINE__, _module, _level));\
+  }else{\
+    getLogInstance()->log(new LogMsg(std::string(log_data), __FILE__, __LINE__, _module, _level));\
+  }\
+  free(log_data);\
+}while(0)
+
+#else
+#define DEBUG_LOG_PRINT(_level,_module, ...) //std::printf(__VA_ARGS__)
```

Comment:
done nothing?

Reply:
yes, save log to file if(DEBUG_MODE && DEBUG_LOG_TO_FILE), else do nothing